File: packages/agent/src/graphs/creativeGraph.ts
import type { Material, Product, RenderPlan, Scene, Script, GenerationTrace } from "@advivid/shared";
import { creativeNodes, type CreativeAgentState } from "../nodes/mockNodes.js";

export interface RunCreativeAgentInput {
  projectId: string;
  scriptId: string;
  product: Product;
  materials: Material[];
}

export interface RunCreativeAgentResult {
  script: Script;
  scenes: Scene[];
  renderPlan: RenderPlan;
  trace: GenerationTrace[];
}

async function runSerial(input: RunCreativeAgentInput): Promise<CreativeAgentState> {
  let state: CreativeAgentState = {
    ...input,
    trace: []
  };

  for (const node of [
    creativeNodes.productAnalyzer,
    creativeNodes.materialRetriever,
    creativeNodes.strategySelector,
    creativeNodes.scriptWriter,
    creativeNodes.scenePlanner,
    creativeNodes.reviewAgent,
    creativeNodes.renderPlanner
  ]) {
    const update = await node(state);
    state = { ...state, ...update };
  }

  return state;
}

async function runWithLangGraph(input: RunCreativeAgentInput): Promise<CreativeAgentState> {
  const mod = await import("@langchain/langgraph");
  const Annotation = mod.Annotation as any;
  const StateGraph = mod.StateGraph as any;
  const START = mod.START as string;
  const END = mod.END as string;

  const State = Annotation.Root({
    projectId: Annotation(),
    scriptId: Annotation(),
    product: Annotation(),
    materials: Annotation(),
    productProfile: Annotation(),
    retrievedMaterials: Annotation(),
    creativeStrategy: Annotation(),
    scriptDraft: Annotation(),
    scenes: Annotation(),
    reviewResult: Annotation(),
    renderPlan: Annotation(),
    trace: Annotation({
      reducer: (_left: GenerationTrace[], right: GenerationTrace[]) => right,
      default: () => []
    })
  });

  const graph = new StateGraph(State)
    .addNode("ProductAnalyzer", creativeNodes.productAnalyzer)
    .addNode("MaterialRetriever", creativeNodes.materialRetriever)
    .addNode("StrategySelector", creativeNodes.strategySelector)
    .addNode("ScriptWriter", creativeNodes.scriptWriter)
    .addNode("ScenePlanner", creativeNodes.scenePlanner)
    .addNode("ReviewAgent", creativeNodes.reviewAgent)
    .addNode("RenderPlanner", creativeNodes.renderPlanner)
    .addEdge(START, "ProductAnalyzer")
    .addEdge("ProductAnalyzer", "MaterialRetriever")
    .addEdge("MaterialRetriever", "StrategySelector")
    .addEdge("StrategySelector", "ScriptWriter")
    .addEdge("ScriptWriter", "ScenePlanner")
    .addEdge("ScenePlanner", "ReviewAgent")
    .addEdge("ReviewAgent", "RenderPlanner")
    .addEdge("RenderPlanner", END)
    .compile();

  return graph.invoke({
    ...input,
    trace: []
  });
}

export async function runCreativeAgent(input: RunCreativeAgentInput): Promise<RunCreativeAgentResult> {
  let state: CreativeAgentState;

  try {
    state = await runWithLangGraph(input);
  } catch {
    state = await runSerial(input);
  }

  if (!state.scriptDraft || !state.scenes || !state.renderPlan) {
    throw new Error("Creative agent did not produce a complete script.");
  }

  return {
    script: {
      ...state.scriptDraft,
      scenes: state.scenes
    },
    scenes: state.scenes,
    renderPlan: state.renderPlan,
    trace: state.trace
  };
}

File: apps/agent-python/app/graphs/creative_graph.py
from __future__ import annotations

from langgraph.graph import END, START, StateGraph

from app.nodes import (
    CreativeState,
    material_retriever,
    product_analyzer,
    render_planner,
    review_agent,
    scene_planner,
    script_writer,
    strategy_selector,
)
from app.schemas import GenerateScriptRequest, GenerateScriptResponse, Script


def build_creative_graph():
    graph = StateGraph(CreativeState)
    graph.add_node("ProductAnalyzer", product_analyzer)
    graph.add_node("MaterialRetriever", material_retriever)
    graph.add_node("StrategySelector", strategy_selector)
    graph.add_node("ScriptWriter", script_writer)
    graph.add_node("ScenePlanner", scene_planner)
    graph.add_node("ReviewAgent", review_agent)
    graph.add_node("RenderPlanner", render_planner)

    graph.add_edge(START, "ProductAnalyzer")
    graph.add_edge("ProductAnalyzer", "MaterialRetriever")
    graph.add_edge("MaterialRetriever", "StrategySelector")
    graph.add_edge("StrategySelector", "ScriptWriter")
    graph.add_edge("ScriptWriter", "ScenePlanner")
    graph.add_edge("ScenePlanner", "ReviewAgent")
    graph.add_edge("ReviewAgent", "RenderPlanner")
    graph.add_edge("RenderPlanner", END)
    return graph.compile()


CREATIVE_GRAPH = build_creative_graph()


async def run_creative_graph(request: GenerateScriptRequest) -> GenerateScriptResponse:
    state = await CREATIVE_GRAPH.ainvoke(
        {
            "projectId": request.projectId,
            "scriptId": request.scriptId,
            "product": request.product,
            "materials": request.materials,
            "trace": [],
        }
    )

    scenes = state["scenes"]
    script = Script(**state["scriptDraft"], scenes=scenes)
    return GenerateScriptResponse(
        script=script,
        scenes=scenes,
        renderPlan=state["renderPlan"],
        trace=state["trace"],
    )
